[#73] no write access fail gracefully

Eclipse-Dominator opened the pull request:
Fixes #73 
Continued from #77 

When there is insufficient permission to write data to the files, an `AccessDeniedException` will appear in the error message. 
This is not as informative as the error message is the same for both `AccessDeniedException` and `IOException`. Thus let's make the error message more descriptive for both exceptions.

Besides this, the same error message (`Invalid Command: <command text>` is logged whenever a CommandException or ParseException occurs. However, such errors can be due to IOExceptions which is not related to the invalid command. 

Therefore, there is also a need to improve the accuracy of the error message by changing it to more accurately describe the failure.

Eclipse-Dominator commented:
Apologies for all the changes, I realised I forgot to include a line spacing between the commit header and the commit body

damithc commented:
> There is an issue with the exit command where exit will attempt to save all file before exiting and due to the permission error, it will prevent the program from exiting.

hmm... there is no real need to save data upon exit. Perhaps we can skip the data saving if the command is the exit command?

Eclipse-Dominator commented:
Hmm I don't think that there is a very clean way to identify specific command without hardcoding certain checks. 

```java
@Override
    public CommandResult execute(String commandText) throws CommandException, ParseException {
        logger.info("----------------[USER COMMAND][" + commandText + "]");

        CommandResult commandResult;
        Command command = addressBookParser.parseCommand(commandText);
        commandResult = command.execute(model);

        try {
            storage.saveAddressBook(model.getAddressBook());
        } catch (IOException ioe) {
            throw new CommandException(FILE_OPS_ERROR_MESSAGE + ioe, ioe);
        }

        return commandResult;
    }
```

```java
    private CommandResult executeCommand(String commandText) throws CommandException, ParseException {
        try {
            CommandResult commandResult = logic.execute(commandText);
            logger.info("Result: " + commandResult.getFeedbackToUser());
            resultDisplay.setFeedbackToUser(commandResult.getFeedbackToUser());

            if (commandResult.isShowHelp()) {
                handleHelp();
            }

            if (commandResult.isExit()) {
                handleExit();
            }

            return commandResult;
        } catch (CommandException | ParseException e) {
            logger.info("Invalid command: " + commandText);
            resultDisplay.setFeedbackToUser(e.getMessage());
            throw e;
        }
    }
```
Saving of addressbook is done in execute itself and only afterwards verification of whether something is help or exit is done. 

---
I was thinking of allowing execute to throw a custom exception to represent critical error where the program need to exit instead of a CommandException
```java
        try {
            storage.saveAddressBook(model.getAddressBook());
        } catch (IOException ioe) {
            throw new CriticalException(FILE_OPS_ERROR_MESSAGE + ioe, ioe); // <-- 
        }
```
Then specifically catching it in `MainWindow.java`
```java
        catch (CriticalException e) {
            handleCriticalError(e); // <-- spawns a popup that exits program after closing. 
        }
```
---
Alternatively, we can also store the exception in CommandResult instead of catching it later on. In a sense we are decoupling the save exception from the parse exceptions. However, both of these method could be an overkill to handle permission denied error.

damithc commented:
As this is a very rare case, we should try a solution that minimizes changes to the code. So, approach 3 should be fine.
Do you think this PR can be just one commit?

Eclipse-Dominator commented:
I think 2 commit is probably more suited here?
1. Decoupling IOException and AccessDeniedException
2. Decoupling ParseException and CommandException

As for the commit body, I am thinking of 

1.  
```
Better error message when file have no write-access

When the data file has no write access, a message containing.
a raw exception is displayed to the user with no specific instructions
on handling such errors.

Let's 
* Polish up error messages displayed to the user
* Decouple IOExceptions and AccessDeniedExceptions
* Add instructions to ask the users to manually close AB3
```
2. 
```
Decoupling of ParseException and CommandException log messages

All errors related to command is logged as "invalid command".
Let's be more specific by decoupling ParseExceptions and CommandExceptions
to enable an more accurate logging of the errors.
```

damithc commented:
I'm not sure if the second commit adds enough value. While the same prefix is used in both cases, the error message itself should tell us what exactly happened, right?
Don't forget that commit message subject should be in imperative mood.

Eclipse-Dominator commented:
I have updated the commit to the following:
```
Improve error message if no write permission

Insufficient permission to read the data file will show
an AccessDeniedException as error message.
AB3 should be able to fail gracefully with instructions
to close the application upon failure.

Let's
- make the error message more user friendly by informing
the user that there are insufficient permissions for the file.
- Decouple AccessDeniedException from IOException messages.
- Decouple ParseException and CommandException log messages.
```

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: %s\n"
+            + "Please exit the application immediately via the exit button";
```
Is there a better alternative to 'via the exit button'? The reader might wonder "what's the 'exit button'?"

damithc replied:
Exiting doesn't solve the problem. We can inform user how to work around the error.

Eclipse-Dominator replied:
Maybe we can use 'X' button? I don't think there is a proper term that is easily understandable for the exit button.

I have included the following "work arounds":
            "1. Running the application as an administrator.\n"
            "2. Changing the permissions of the save files or folder to allow read write access.\n"
            "3. Changing the save location of the application to a location that you have read write access to.\n";

damithc replied:
> Maybe we can use 'X' button? I don't think there is a proper term that is easily understandable for the exit button.

Yes that might work.  e.g., `exit the application by clicking the eXit button (i.e., the button marked with an 'X' icon)`


>3\. Changing the save location of the application to a location that you have read write access to.

The user might wonder how to change the save location.

Perhaps we can just tell the user the _intent_ of the workaround and let her figure out a mechanism? e.g., `Before running it again, ensure the application .jar file has write permissions to the folder it is in.`

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -47,8 +52,12 @@ public CommandResult execute(String commandText) throws CommandException, ParseE
         try {
             storage.saveAddressBook(model.getAddressBook());
+        } catch (AccessDeniedException e) {
+            String message = e.getMessage();
```
inline this, as it is used only once?

damithc commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -187,10 +187,14 @@ private CommandResult executeCommand(String commandText) throws CommandException
+            logger.info("An error occured while executing this command: " + commandText);
+            resultDisplay.setFeedbackToUser(e.getMessage());
+            throw e;
         }
```
I don't think this change is related to the issue at hand. It should be a separate PR.
Also, I think the following tweak is enough (no need for two catch blocks).
```suggestion
        } catch (CommandException | ParseException e) {
            logger.info("An error occured while executing this command: " + commandText);
            resultDisplay.setFeedbackToUser(e.getMessage());
            throw e;
        }
```

Eclipse-Dominator replied:
I separated the 2 because in the log file, any command that triggered CommandException or ParseException previously will log the line "Invalid Command: %s", even though the command is valid.

Perhaps we can standardise to a general term "An error occured while executing this command: " instead of invalid command as suggested?

damithc replied:
>any command that triggered CommandException or ParseException previously will log the line "Invalid Command: %s", even though the command is valid.

I agree, that doesn't sound right.

> Perhaps we can standardise to a general term "An error occured while executing this command: " instead of invalid command as suggested?

Yes, that's what I'm suggesting as well.

damithc reviewed (changes requested):
Some suggestions added.

Commit message: Use 'application' instead of 'AB3' as other higher AB versions (e.g., AB4) are derived from this code base as well.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,13 @@
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: %s\n"
+            + "Please exit the application immediately via the exit button.";
```
Probably no need for the word `immediately`?

damithc replied:
`via` -> `by clicking`

damithc replied:
Why does this message asks the user to exit the application but doesn't say how to recover?

Eclipse-Dominator replied:
This error message is for generic IOException errors that are not AccessDeniedExceptions, so I am not sure what the standard procedure is for recovering from such errors, so I left it as such.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,13 @@
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: %s\n"
+            + "Please exit the application immediately via the exit button.";
+    public static final String FILE_OPS_PERM_ERROR_MESSAGE = "Could not save data to file: "
```
`PERM` is not a standard abbreviation. Use the full word.

damithc reviewed (changes requested):
Commit message: 
* subject says `write`, first sentence says `read`
* what we are doing is making the error message more informative (rather than 'fail gracefully), right? If so, we can omit the phrase 'fail gracefully'.
* I don't think we are doing any intentional 'decoupling' in this PR. It's just incidental that we add a new exception branch.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,13 @@
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: %s\n"
+            + "Please exit the application by clicking the exit button (i.e., the button marked with an 'X' icon).";
```
I still think either you have to tell the user how to recover, or just state the problem only (i.e., without mention exiting at all).

Eclipse-Dominator replied:
I think I'll state the error and mention continued usage might lead to data loss

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,13 @@
+            + "Please exit the application by clicking the exit button (i.e., the button marked with an 'X' icon).";
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE = "Could not save data to file: "
+            + "Insufficient Permissions for %s.\n\n"
+            + "Please exit the application by clicking the exit button (i.e., the button marked with an 'X' icon).\n"
```
It's been a while so I might have forgotten the reason; why the user has to exit the application?

Eclipse-Dominator replied:
Not exiting the application is fine but none of the changes will be saved until the permissions is changed. Perhaps we can instruct the user to change the permissions of the relevant files/folder on the go so that future commands will be saved but there could be complications so it might be better to just exit the application straight away and change the permission rather than changing the permissions on the go

Exiting via the x button is required as typing exit will not cause the application to quit

Eclipse-Dominator replied:
I have updated the message to 
```
Could not save data to file: Insufficient permissions for FILENAME

Please ensure the application .jar file has write permissions to the folder it is in.
Continued usage of the application without fixing this error might lead to data loss.
You can exit the application by clicking the exit button (i.e., the button marked with an 'X' icon).
```

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,13 @@
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: %s\n"
+            + "Please exit the application by clicking the exit button (i.e., the button marked with an 'X' icon).";
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE = "Could not save data to file: "
+            + "Insufficient Permissions for %s.\n\n"
```
Permissions -> permissions?

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -84,7 +86,28 @@ public void execute_storageThrowsIoException_throwsCommandException() {
+        String expectedMessage =
+                String.format(LogicManager.FILE_OPS_PERMISSION_ERROR_MESSAGE, DUMMY_AD_EXCEPTION.getMessage());
         assertCommandFailure(addCommand, CommandException.class, expectedMessage, expectedModel);
     }
```
Good job adding a test. 
Looks almost identical to an existing test method. Perhaps the common code can be extracted out as a utility method?

damithc reviewed (changes requested):
Some comments on a few things that I didn't think of before.

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +163,19 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
 
     /**
-     * A stub class to throw an {@code IOException} when the save method is called.
+     * A method that returns a {@code StorageManager} with the {@code JsonAddressBookStorage} that
```
This is not how we phrase the overview statement of methods.

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +163,19 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+        };
+        JsonUserPrefsStorage userPrefsStorage =
+                new JsonUserPrefsStorage(temporaryFolder.resolve("ExceptionUserPrefs.json"));
+        return new StorageManager(addressBookStorage, userPrefsStorage);
```
Seems not worth extracting this. The extra long name is also an indication of extra complexity -- the v5 stub class is easier to understand than this stub-creating-method. 

How about something like this (because the only difference between the two test cases is the exception and the expected error message)?
```java
    @Test
    public void execute_storageThrowsIoException_throwsCommandException() {
        assertCommandFailureForException(DUMMY_AD_EXCEPTION, 
            String.format(LogicManager.FILE_OPS_PERMISSION_ERROR_MESSAGE, DUMMY_AD_EXCEPTION.getMessage());
    }
    
    @Test
    public void execute_storageThrowsIoException_throwsCommandException() {
        assertCommandFailureForException(DUMMY_IO_EXCEPTION,
            String.format(LogicManager.FILE_OPS_ERROR_MESSAGE, DUMMY_IO_EXCEPTION.getMessage()));
    }
```

Alternatively, we can go back to the v5 test cases with the duplicated code. We can generalize further when we have a third similar test case.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,14 @@
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data! The following errors has occured: %s\n"
+            + "Continued usage of the application without fixing this error might lead to data loss.\n";
```
Let's remove this line as well. Reasons: 1. it seems like an obvious thing. 2. it's talking about _consequence_ of the error from an overall application's perspective. It's cleaner to stick with just the description of the error only.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,14 @@
  */
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data! The following errors has occured: %s\n"
```
Unless this error is more severe than the other one, no good reason to use an `!` but not the other.  The phrasing of the error message can be same (for consistency) except there is a good reason to differ.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,12 @@
+
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE = "Could not save data to file: "
+            + "Insufficient permissions to write content to %s.\n\n"
+            + "The application currently do not have write permissions to the folder it is in.\n";
```
Are we able to say this for sure? What if the file exists but is read-only? 
Also, the fact that the data file is created in the home folder a separate piece of information that might change independent of this code. So, I don't think it is safe to say something like this here. What do you think?

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+     * - the resulting error message is equal to {@code expectedMessage} <br>
+     *
+     * @param e exception to be thrown by the AddressBookStorage
+     * @param expectedMessage expected message to be thrown
```
Execute -> Executes

Explain WHAT the method does rather than HOW it does it.

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+                    throws IOException {
+                throw e;
+            }
+        };
```
Nice way to avoid an extra class.  👍

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
-        public void saveAddressBook(ReadOnlyAddressBook addressBook, Path filePath) throws IOException {
-            throw DUMMY_IO_EXCEPTION;
-        }
+    private void assertCommandFailureForException(IOException e, String expectedMessage) {
```
Make the name a bit more specific?
```suggestion
    private void assertCommandFailureForExceptionFromStorage(IOException e, String expectedMessage) {
```

damithc reviewed (changes requested):
Some nits ...

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following errors has occured: %s\n";
+
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE = "Could not save data to file: "
+            + "Insufficient permissions to write content to file %s.\n";
```
```suggestion
    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE = "Could not save data to file. "
            + "Insufficient permissions to write to the file %s.\n";
```
* Made punctuation more consistent.
* Removed redundant word `content`

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -33,7 +34,8 @@
 public class LogicManagerTest {
-    private static final IOException DUMMY_IO_EXCEPTION = new IOException("dummy exception");
+    private static final IOException DUMMY_IO_EXCEPTION = new IOException("dummy io exception");
+    private static final IOException DUMMY_AD_EXCEPTION = new AccessDeniedException("dummy ad exception");
```
```suggestion
    private static final IOException DUMMY_IO_EXCEPTION = new IOException("dummy IO exception");
    private static final IOException DUMMY_AD_EXCEPTION = new AccessDeniedException("dummy access denied exception");
```

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -70,22 +72,14 @@ public void execute_validCommand_success() throws Exception {
+    @Test
+    public void execute_storageThrowsAdException_throwsCommandException() {
+        assertCommandFailureForExceptionFromStorage(DUMMY_AD_EXCEPTION, String.format(
+                LogicManager.FILE_OPS_PERMISSION_ERROR_MESSAGE, DUMMY_AD_EXCEPTION.getMessage()));
```
Inconsistent line wrapping compared to line 75 (`String` vs `String.format(`)

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+     * Test the handling of an {@code IOException} in a dummy command execution scenario by the
+     * Storage component. This test verifies that:<br>
+     * - The expected exception ({@code e}) is thrown<br>
+     * - The resulting error message matches the expected message ({@code expectedMessage})<br>
```
```suggestion
     * Tests the Logic component's handling of an {@code IOException} thrown by 
     *  the Storage component. 
```

* Test -> Tests
* Shortened the description, as test descriptions need not be very detailed. The previous details were not complete anyway, as it did not mention about verifying the Model component is as expected.

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+     * - The resulting error message matches the expected message ({@code expectedMessage})<br>
+     *
+     * @param e the exception to be thrown by the AddressBookStorage
+     * @param expectedMessage the expected message to be thrown
```
```suggestion
     * @param e the exception to be thrown by the Storage component 
     * @param expectedMessage the message expected inside exception thrown by the Logic component
```

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+    private void assertCommandFailureForExceptionFromStorage(IOException e, String expectedMessage) {
+        Path prefPath = temporaryFolder.resolve("ExceptionUserPrefs.json");
+
+        // Setup LogicManager with an AddressBookStorage that throws an IOException when saving
```
```suggestion
        // Inject LogicManager with an AddressBookStorage that throws the IOException e when saving
```

damithc commented on src/test/java/seedu/address/logic/LogicManagerTest.java:
```diff
@@ -147,16 +141,38 @@ private void assertCommandFailure(String inputCommand, Class<? extends Throwable
+
+        logic = new LogicManager(model, storage);
+
+        // Execute add command
```
```suggestion
        // To trigger the save method, execute an add command
```

damithc commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,7 +188,7 @@ private CommandResult executeCommand(String commandText) throws CommandException
             return commandResult;
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
+            logger.info("An error occured while executing this command: " + commandText);
```
Should this change be in a separate commit, as it is not related to other changes in this commit?

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -47,8 +52,10 @@ public CommandResult execute(String commandText) throws CommandException, ParseE
         } catch (IOException ioe) {
-            throw new CommandException(FILE_OPS_ERROR_MESSAGE + ioe, ioe);
+            throw new CommandException(String.format(FILE_OPS_ERROR_MESSAGE, ioe.getMessage()), ioe);
         }
```
@Eclipse-Dominator What kind of problems would cause IOException to be thrown instead of AccessDeniedException ?

Eclipse-Dominator replied:
There are some exceptions like DiskFullException that could be thrown. Generally it is part of a niche and unlikely scenarios that will cause an IOException to be thrown so this catch block is generally used to handle them.

damithc reviewed (changes requested):
Added some stylistic suggestions. Feel free to deviate if you think of better alternatives.

Might need to change the commit message to reflect the current version of the code too.

Eclipse-Dominator committed:
Improve error message if no write permission

Insufficient permission to write to the data file will show up as
AccessDeniedException in the error message.
The application should be able to describe the error to the user rather
than just showing the name of the Exception.

Let's
- make the error message more descriptive by informing
the user that there are insufficient permissions for the file for
AccesssDeniedExceptions.

- Also include a better general purpose description for other forms
of IO related exceptions

Eclipse-Dominator committed:
Improve logs msg for command related exception

Logger will log "Invaid Command: <command text>" when a CommandException
or ParseException occur.

However, the exceptions might be caused by other issuess like unable to
write to save file etc.
Therefore, the current log message is not suitable for CommandException.

Let's improve the accuracy of the error message by changing it to
accurately describe both CommandException and ParseException.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following errors has occured: %s\n";
+
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE =
+            "Could not save data to file. Insufficient permissions to write to the file %s.\n";
```
`to the file (or the folder):`?

damithc reviewed (commented):
Commit 2: good. It's a tiny change but definite value added.
Commit 1: Feels like the value added is marginal (and code to value ratio is very low), but I'm OK to keep it. OK to drop it as well. The low value of the commit is not your fault -- it's because I changed my mind about giving recovery instructions to the user.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following errors has occured: %s\n";
+
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE =
+            "Could not save data to file (or the folder). Insufficient permissions to write to the file %s.\n";
```
```suggestion
    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE =
            "Could not save data to file %s. Insufficient permissions to write to the file (or the folder).\n";
```

damithc reviewed (changes requested):
Cosmetic suggestion.
Also, update the PR description to match the current state so that we can ask others to review this PR.

damithc commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
+
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE =
+            "Could not save data to file %s. Insufficient permissions to write to the file (or the folder).\n";
+
```
Just noticed the `\n` at the end. Doesn't seem to belong there, as the error message is a one-liner.

damithc reviewed (commented):
Minor cosmetic issue.

damithc commented:
@se-edu/tech-team-level1 for your review ...

chia-yh commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
  */
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following errors has occured: %s\n";
```
```suggestion
    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following error has occured: %s.\n";
```

chia-yh replied:
Not sure if the comment about `\n` at the end for `FILE_OPS_PERMISSION_ERROR_MESSAGE` also applies here

chia-yh reviewed (commented):
Some minor nits about the error message, but otherwise looks fine. I like the change to use `String.format` for the error message, as it feels cleaner. I also think it's an excellent idea to have the error message be more informative/accurate.

SPWwj reviewed (commented):
<img width="643" alt="image" src="https://github.com/se-edu/addressbook-level3/assets/30100720/4e6bba22-04a0-434c-aa16-ac3b25424e1a">

Just a minor sugguestion on the user interface. Currently, the exception messages are too long, causing some readability issues. It may be more user-friendly to insert line breaks into these messages to enhance their legibility.

Other than this, everything look good!

damithc commented:
> Just a minor sugguestion on the user interface. Currently, the exception messages are too long, causing some readability issues. It may be more user-friendly to insert line breaks into these messages to enhance their legibility.

Fair point @SPWwj Probably OK in this case because the second sentence follows directly from the first. We don't want to give the impression that it is a separate problem. In fact I wonder if we should phrase it as `...book.json due to insufficient ...` 🤔

SPWwj commented:
> > Just a minor sugguestion on the user interface. Currently, the exception messages are too long, causing some readability issues. It may be more user-friendly to insert line breaks into these messages to enhance their legibility.
> 
> Fair point @SPWwj Probably OK in this case because the second sentence follows directly from the first. We don't want to give the impression that it is a separate problem. In fact I wonder if we should phrase it as `...book.json due to insufficient ...` 🤔

Agreed, this completes the sentence, enhancing clarity and precision by providing a reason. In circumstances with multiple issues, it ensures that no ambiguity arises from multiple sentences.

SPWwj commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following errors has occured: %s";
+
+    public static final String FILE_OPS_PERMISSION_ERROR_MESSAGE =
+            "Could not save data to file %s due to insufficient permissions to write to the file (or the folder).\n";
```
```suggestion
            "Could not save data to file %s due to insufficient permissions to write to the file or the folder.";
```

1. My personal opinion is that components using the strings should handle additional formatting elements, such as newline characters. I invite further discussion on this.
2. In my view, the phrase "(or the folder)" is redundant. Although it is commonly seen, it doesn't offer any differentiation from "or the folder," which is two characters fewer and provides a cleaner presentation.

Eclipse-Dominator replied:
AB3 right now does not really support composing messages, especially exceptions, the closest element that is related to this is `MainWindow` where the related method is shown below:

```java
/**
     * Executes the command and returns the result.
     *
     * @see seedu.address.logic.Logic#execute(String)
     */
    private CommandResult executeCommand(String commandText) throws CommandException, ParseException {
        try {
            CommandResult commandResult = logic.execute(commandText);
            logger.info("Result: " + commandResult.getFeedbackToUser());
            resultDisplay.setFeedbackToUser(commandResult.getFeedbackToUser());

            if (commandResult.isShowHelp()) {
                handleHelp();
            }

            if (commandResult.isExit()) {
                handleExit();
            }

            return commandResult;
        } catch (CommandException | ParseException e) {
            logger.info("An error occured while executing this command: " + commandText);
            resultDisplay.setFeedbackToUser(e.getMessage());
            throw e;
        }
    }
```

imo, the scenario right now is quite literally what you see is what you get for the handling of exceptions. For the `Exception`s  handled in this PR, the component handling and using the string is `LogicManager` So I think it is ok to remove the newline character and keep the handling as it is.


As for 2, I do agree that "(or the folder)" feels redundant but it is also possible to change the folder permission to readonly so it can be slightly different compared to the file. However, ultimately it is the file that does not have read permission due to the folder settings. 

I am okay with removing the phrase but I feel that adding it can be more accurate to the exception.

damithc replied:
1. I agree that `\n` should be removed. In fact I believe it was left there by mistake. 
2. I think @SPWwj is asking to remove the parenthesis, not the whole phrase.

hansstanley commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
  */
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following errors has occured: %s";
```
Just some minor grammar changes, assuming that only one error follows this message:

```suggestion
    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data. The following error has occurred: %s";
```

To ensure this format string is not used as-is instead of being wrapped in `String.format`, it might also be helpful to rename it to distinguish it from other errors messages, for e.g. `FILE_OPS_ERROR_TEMPLATE` or `FILE_OPS_ERROR_PREFIX`.

Eclipse-Dominator replied:
Good catch, I will follow the formatting name of the classes here!

damithc commented:
@chia-yh @hansstanley @SPWwj Can we have another review, as the code has changed since your last review?

SPWwj commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
  */
 public class LogicManager implements Logic {
-    public static final String FILE_OPS_ERROR_MESSAGE = "Could not save data to file: ";
+    public static final String FILE_OPS_ERROR_FORMAT = "Could not save data. The following error has occurred: %s";
```
```suggestion
    public static final String FILE_OPS_ERROR_FORMAT = "Could not save data due to the following occurred error: %s.";
```

damithc replied:
Perhaps remove `occurred` as well? i.e., `Could not save data due to the following error`

chia-yh reviewed (commented):
LGTM, though I also think the suggested changes to `FILE_OPS_ERROR_FORMAT` are a good idea as it reads better.

hansstanley reviewed (commented):
Other than the existing comments, LGTM

SPWwj commented on src/main/java/seedu/address/logic/LogicManager.java:
```diff
@@ -21,7 +22,11 @@
+    public static final String FILE_OPS_ERROR_FORMAT = "Could not save data due to the following error: %s";
+
+    public static final String FILE_OPS_PERMISSION_ERROR_FORMAT =
+            "Could not save data to file %s due to insufficient permissions to write to the file or the folder.";
```
```suggestion
            "Could not perform write operations to file %s due to insufficient permissions";
```

damithc replied:
I wouldn't mind keeping the original, as it is more informative and context-specific.

Eclipse-Dominator replied:
I think the original feels more specific to the actual warning as well, `perform write operations` might be a bit too generic.

SPWwj commented on src/main/java/seedu/address/ui/MainWindow.java:
```diff
@@ -188,7 +188,7 @@ private CommandResult executeCommand(String commandText) throws CommandException
             return commandResult;
         } catch (CommandException | ParseException e) {
-            logger.info("Invalid command: " + commandText);
+            logger.info("An error occured while executing this command: " + commandText);
```
```suggestion
            logger.info("An error occurred while executing command:: " + commandText);
```

damithc replied:
@SPWwj Should be only one colon?

SPWwj replied:
Oh yes 😂

```suggestion
            logger.info("An error occurred while executing command: " + commandText);
```

damithc reviewed (approved):
No further comments from me. LGTM. 👍

damithc commented:
Thanks for these improvements, @Eclipse-Dominator
